src/couch/src/couch_secrets.erl

```diff
@@ -0,0 +1,193 @@
+    update_all_secrets(),
+    erlang:send_after(5000, self(), cache_cleanup),
+    ok = config:listen_for_changes(?MODULE, undefined),
+    {ok, nil, {continue, get_secrets}}.
+
+handle_call({insert, {Node, current}, Secret}, _From, State) ->
+    case current_secret_from_ets(Node) of
+        undefined ->
+            ets:insert(?MODULE, [{{Node, current}, Secret}]);
+        OldSecret ->
+            TimeoutSecs = chttpd_util:get_chttpd_auth_config_integer("timeout", 600),
+            ExpiresAt = erlang:system_time(second) + TimeoutSecs,
+            ets:insert(?MODULE, [{{Node, current}, Secret}, {{Node, ExpiresAt}, OldSecret}])
+    end,
+    update_all_secrets(),
+    {reply, ok, State};
+handle_call({insert, Key, Secret}, _From, State) ->
+    ets:insert(?MODULE, {Key, Secret}),
+    update_all_secrets(),
+    {reply, ok, State};
+handle_call(get_secrets, _From, State) ->
+    Secrets = ets:match_object(?MODULE, {{node(), '_'}, '_'}),
+    {reply, Secrets, State};
+handle_call(flush_cache, _From, State) ->
+    %% used from tests to prevent spurious failures due to timing
+    MatchSpec = [{{{'_', '$1'}, '_'}, [{is_integer, '$1'}], [true]}],
+    NumDeleted = ets:select_delete(?MODULE, MatchSpec),
+    if
+        NumDeleted > 0 -> update_all_secrets();
+        true -> ok
+    end,
+    {reply, NumDeleted, State};
+handle_call(_Msg, _From, State) ->
+    {noreply, State}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_continue(get_secrets, State) ->
+    {Replies, _BadNodes} = gen_server:multi_call(nodes(), ?MODULE, get_secrets),
+    {_Nodes, Secrets} = lists:unzip(Replies),
+    true = ets:insert(?MODULE, lists:flatten(Secrets)),
+    update_all_secrets(),
+    {noreply, State}.
+
+handle_info(restart_config_listener, State) ->
+    ok = config:listen_for_changes(?MODULE, nil),
+    update_current_secret(),
+    {noreply, State};
+handle_info(cache_cleanup, State) ->
+    erlang:send_after(5000, self(), cache_cleanup),
+    Now = os:system_time(second),
+    MatchSpec = [{{{'_', '$1'}, '_'}, [{is_integer, '$1'}, {'<', '$1', Now}], [true]}],
+    NumDeleted = ets:select_delete(?MODULE, MatchSpec),
+    if
+        NumDeleted > 0 -> update_all_secrets();
+        true -> ok
+    end,
+    {noreply, State};
+handle_info(_Msg, State) ->
+    {noreply, State}.
+
+handle_config_change("chttpd_auth", "secret", _, _, _) ->
+    update_current_secret(),
+    {ok, undefined};
+handle_config_change("couch_httpd_auth", "secret", _, _, _) ->
+    update_current_secret(),
+    {ok, undefined};
+handle_config_change(_, _, _, _, _) ->
+    {ok, undefined}.
+
+handle_config_terminate(_, stop, _) ->
+    ok;
+handle_config_terminate(_Server, _Reason, _State) ->
+    erlang:send_after(3000, whereis(?MODULE), restart_config_listener).
+
+%% private functions
+
+update_current_secret() ->
+    NewSecret = current_secret_from_config(),
+    spawn_link(fun() ->
```

Comment:
This will link to the config gen_event callback, wonder if it's better to just send a message to the couch_secrets gen_server and let that do the linking and spawning the mult-call process

Reply:
removed